src/runtime-rs/crates/resource/src/cpu_mem/cpu.rs

```diff
@@ -0,0 +1,166 @@
+
+        let curr_vcpus = self
+            .do_update_cpu_resources(vcpu_required, op, hypervisor, agent)
+            .await?;
+        self.update_current_vcpu(curr_vcpus).await;
+        Ok(())
+    }
+
+    async fn current_vcpu(&self) -> u32 {
+        let current_vcpu = self.current_vcpu.read().await;
+        *current_vcpu
+    }
+
+    async fn update_current_vcpu(&self, new_vcpus: u32) {
+        let mut current_vcpu = self.current_vcpu.write().await;
+        *current_vcpu = new_vcpus;
+    }
+
+    // update container_cpu_resources field
+    async fn update_container_cpu_resources(
+        &self,
+        cid: &str,
+        linux_resources: Option<&LinuxResources>,
+        op: ResourceUpdateOp,
+    ) -> Result<()> {
+        if let Some(resources) = linux_resources {
+            if let Some(cpu) = &resources.cpu {
+                let container_resource = LinuxContainerCpuResources::try_from(cpu)?;
+                let mut resources = self.container_cpu_resources.write().await;
+                match op {
+                    ResourceUpdateOp::Add | ResourceUpdateOp::Update => {
+                        resources.insert(cid.to_owned(), container_resource);
+                    }
+                    ResourceUpdateOp::Del => {
+                        resources.remove(cid);
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    // calculates the totoal required vcpus by adding each containers' requirements within the pod
+    async fn calc_cpu_resources(&self) -> Result<u32> {
+        let mut total_vcpu = 0;
+        let mut cpuset_vcpu: HashSet<u32> = HashSet::new();
+
+        let resources = self.container_cpu_resources.read().await;
+        for (_, cpu_resource) in resources.iter() {
+            let vcpu = cpu_resource.get_vcpus().unwrap_or(0) as u32;
+            cpuset_vcpu.extend(cpu_resource.cpuset().iter());
+            total_vcpu += vcpu;
+        }
+
+        // contrained only by cpuset
+        if total_vcpu == 0 && !cpuset_vcpu.is_empty() {
+            info!(sl!(), "(from cpuset)get vcpus # {:?}", cpuset_vcpu);
+            return Ok(cpuset_vcpu.len() as u32);
+        }
+
+        info!(
+            sl!(),
+            "(from cfs_quota&cfs_period)get vcpus count {}", total_vcpu
+        );
+        Ok(total_vcpu)
+    }
+
+    // do hotplug the vcpu
+    async fn do_update_cpu_resources(
+        &self,
+        new_vcpus: u32,
+        op: ResourceUpdateOp,
+        hypervisor: &dyn Hypervisor,
+        agent: &dyn Agent,
+    ) -> Result<u32> {
+        let old_vcpus = self.current_vcpu().await;
+        if (op == ResourceUpdateOp::Add && old_vcpus > new_vcpus)
+            || (op == ResourceUpdateOp::Del && old_vcpus < new_vcpus)
+        {
+            return Ok(old_vcpus);
+        }
+
+        let (old, new) = hypervisor
+            .resize_vcpu(old_vcpus, new_vcpus)
+            .await
+            .context("resize vcpus")?;
+
+        if old < new {
```

Comment:
I am confusing why we do online only if  old < new, suggest adding some comment for this